src/gluonnlp/model/attention_cell.py

```diff
@@ -482,3 +492,128 @@ def _compute_weight(self, F, query, key, mask=None):
+                                     flatten=False, weight_initializer=weight_initializer,
+                                     bias_initializer=bias_initializer, prefix='key_')
+            self.proj_value = nn.Dense(units=self._value_units, use_bias=self._use_bias,
+                                       flatten=False, weight_initializer=weight_initializer,
+                                       bias_initializer=bias_initializer, prefix='value_')
+            self.key_embedding = nn.Embedding(input_dim=self._max_relative_position, output_dim=self._embed_units)
+            self.value_embedding = nn.Embedding(input_dim=self._max_relative_position, output_dim=self._embed_units)
+
+            self._dropout_layer = nn.Dropout(dropout)
+
+
+    def hybrid_forward(self, F, query, key, value, position_matrix, mask=None):  # pylint: disable=arguments-differ
+        key_embeddings = self.key_embedding(mx.symbol.BlockGrad(position_matrix))
+        value_embeddings = self.value_embedding(mx.symbol.BlockGrad(position_matrix))
+        att_weights = self._compute_weight(F, query, key, key_embeddings, mask)
+        context_vec = self._read_by_weight(F, att_weights, value, value_embeddings)
+        return context_vec, att_weights
+
+    def _compute_weight(self, F, query, key, key_embeddings, mask):
+        query = self.proj_query(query)  # Shape (batch_size, query_length, query_units)
+        # Shape (batch_size * num_heads, query_length, ele_units)
+        query = F.transpose(query.reshape(shape=(0, 0, self._num_heads, -1)),
+                            axes=(0, 2, 1, 3))\
+                 .reshape(shape=(-1, 0, 0), reverse=True)
+        key = self.proj_key(key)
+        key = F.transpose(key.reshape(shape=(0, 0, self._num_heads, -1)),
+                          axes=(0, 2, 1, 3)).reshape(shape=(-1, 0, 0), reverse=True)
+        if mask is not None:
+            mask = F.broadcast_axis(F.expand_dims(mask, axis=1),
+                                    axis=1, size=self._num_heads)\
+                    .reshape(shape=(-1, 0, 0), reverse=True)
+
+        query = F.contrib.div_sqrt_dim(query)
+
+        pos_score = F.batch_dot(query.transpose(axes=(1, 0, 2)), key_embeddings, transpose_b=True)
+
+        att_score = F.batch_dot(query, key, transpose_b=True) + pos_score.transpose(axes=(1, 0, 2))
+
+        att_weights = self._dropout_layer(_masked_softmax(F, att_score, mask, np.float32))
+
+        return att_weights.reshape(shape=(-1, self._num_heads, 0, 0), reverse=True)
+
+    def _read_by_weight(self, F, att_weights, value, value_embeddings):
+        att_weights = att_weights.reshape(shape=(-1, 0, 0), reverse=True)
+        value = self.proj_value(value)
+        value = F.transpose(value.reshape(shape=(0, 0, self._num_heads, -1)),
+                            axes=(0, 2, 1, 3)).reshape(shape=(-1, 0, 0), reverse=True)
+
+        pos_score = F.batch_dot(att_weights.transpose(axes=(1, 0, 2)), value_embeddings)
```

Comment:
`transpose_b=True` is not needed? It is used in computing `pos_score = F.batch_dot(query.transpose(axes=(1, 0, 2)), key_embeddings, transpose_b=True)`

Reply:
Why not needed. Let's sync offline.

Reply:
`value_embeddings` has shape `(length, self._embed_units)`. so `transpose_b=True` is need to compute the `batch_dot`. it is similar to line `pos_score = F.batch_dot(query.transpose(axes=(1, 0, 2)), key_embeddings, transpose_b=True)`.